docs/architecture/adr-069-hasher.md

```diff
@@ -0,0 +1,146 @@
+
+## Decision
+TBD
+
+## Detailed Design
+
+### The new hashing struct
+Introduce `Hasher` interface to inherit [hash.Hash](https://pkg.go.dev/hash#Hash) and extend it with the current Tendermint core existing functions. The `Hasher` implementation is required in each supporting hash algorithm.
+
+```go
+package tmhash
+
+type Hasher interface {
+    hash.Hash
+    SumTruncated(bz []byte) []byte
+}
+```
+
+We would like to define the HashType for using in the later struct.
+```go
+type HashType string
+const (
+    SHA256 HashType = "sha256"
+    BLAKE2B HashType = "blake2b"
+    CUSTOM HashType = "custom"
+    // SHA256_SIMD HashType = "sha256_simd"
+    // other hash type if the TM core want to support
+)
+
+```
+
+The `HasherContainer` provide a container to keep the different `Hasher` instance base on the information
+provides when construcing the node.
+
+```go
+type HasherContainer struct {
+    hasherMap map[HashType]Hasher
+    RegisterType(HashType) error
+    GetHasher(HashType) (Hasher, error)
+}
+
+func (hc HasherContainer) RegisterType(ht HashType, customHasher Hasher) error {
+    switch ht {
+        case string(tmHash.SHA256):
+            hasherMap[ht] := sha256.New()
+            return nil
+        case string(tmHash.BLAKE2B):
+            hasherMap[ht] := blake2b.New()
+            return nil
+        case string(tmHash.CUSTOM):
+            hasherMap[ht] := customhash.New(customHasher)
+            return nil
+        default:
+            return errors.New("register an invalid hash type %w", ht) 
+    }
+}
+
+func (hc HasherContainer) GetHasher(ht HashType) Hasher {
+    return hc.hasherMap[ht]
+}
+
+```
+
+### Customizable hashing
+Because of we introduce a customize HashType - `CUSTOM`. The developer will need to implement their customize hasher and register it into the `HasherContainer` by calling `RegisterType(CUSTOM, customHasher)`
+
+```go
+package tmhash
+
+type CustomHash struct {
+    hasher tmHash.Hasher
+}
+
+var _ tmHash.Hasher = (*CustomHash)(nil)
+
+func New(hr tmhash.Hasher) tmhash.Hasher {
+    return &CustomHash{hasher: hr}
+}
+
+func (ch *CustomHash) SumTruncated(bz []byte) []byte {
+	return ch.hasher.SumTruncated(bz)
+}
+
+// other implementation for the hash.Hash functions
+...
+``` 
+
+### Chain parameter settings
```

Comment:
I'm not the biggest fan of making this consensus param, ideally a developer would pick which hasher to use at compile time/before they distribute the binary. I can be convinced otherwise though.

Reply:
Yap, it's not necessary to put the param to into the genesis. But the settings need to be aligned in the whole network. Once decided, it shouldn't be changed unless the protocol needs to be upgraded.

Reply:
I think this parameter should be a genesis param instead of a consensus https://github.com/tendermint/tendermint/issues/6814